### ISSUE TITLE
Question: Is it possible to see a stack trace in luaj

### ISSUE THREAD

**Issue by desb42** [gnosygnu/xowa#401]
I have been making some changes to my version of xowa and seem to have caused an issue with accessing some wikidata
The page I am looking at is
it.wikipedia.org/wiki/Airbus_Group
I am getting errors of the form

    invoke failed: Airbus_Group {{#invoke:Controllo Wikidata|main}} [err 0] <gplx> =Modulo:Wikidata:721 Tipo di dato sconosciuto
      stack traceback:
      Module:Controllo Wikidata:32: in function <Module:Controllo Wikidata:31>
      mw.lua:531: in function <mw.lua:530>
      [Java]: in function '__index'
      g:/xowa/bin/any/xowa/xtns/Scribunto/engines/Luaj/MWServer.lua:75: in function <g:/xowa/bin/any/xowa/xtns/Scribunto/engines/Luaj/MWServer.lua:74>
      [Java]: in ?: ttl=Airbus_Group excerpt={{#invoke:Controllo Wikidata|main}}
      [trace]:
        gplx.xowa.xtns.scribunto.Scrib_core.Handle_error(Scrib_core.java:216)
        gplx.xowa.xtns.scribunto.engines.luaj.Luaj_engine.Dispatch_as_kv_ary(Luaj_engine.java:80)
        gplx.xowa.xtns.scribunto.engines.luaj.Luaj_engine.CallFunction(Luaj_engine.java:59)
        gplx.xowa.xtns.scribunto.Scrib_core.Invoke(Scrib_core.java:182)
        gplx.xowa.xtns.scribunto.Scrib_invoke_func.Func_evaluate(Scrib_invoke_func.java:69)
        gplx.xowa.parsers.tmpls.Xot_invk_tkn_.Eval_func(Xot_invk_tkn_.java:30)
        gplx.xowa.parsers.tmpls.Xot_invk_tkn.Tmpl_evaluate(Xot_invk_tkn.java:258)
        gplx.xowa.parsers.tmpls.Xot_defn_tmpl.Tmpl_evaluate(Xot_defn_tmpl.java:86)
        gplx.xowa.parsers.tmpls.Xot_invk_tkn.Tmpl_evaluate(Xot_invk_tkn.java:288)
        gplx.xowa.parsers.tmpls.Xot_defn_tmpl.Tmpl_evaluate(Xot_defn_tmpl.java:86)
        gplx.xowa.parsers.tmpls.Xot_invk_tkn.Tmpl_evaluate(Xot_invk_tkn.java:288)
        gplx.xowa.parsers.tmpls.Xot_tmpl_wtr.Write_tkn(Xot_tmpl_wtr.java:103)
        gplx.xowa.parsers.tmpls.Xot_tmpl_wtr.Write_tkn(Xot_tmpl_wtr.java:35)
        gplx.xowa.parsers.tmpls.Xot_tmpl_wtr.Write_all(Xot_tmpl_wtr.java:23)
        gplx.xowa.parsers.Xop_parser.Expand_tmpl(Xop_parser.java:50)
        gplx.xowa.parsers.Xop_parser.Expand_tmpl(Xop_parser.java:44)
        gplx.xowa.parsers.Xop_parser.Parse_text_to_wdom(Xop_parser.java:102)
        gplx.xowa.parsers.Xow_parser_mgr.Parse(Xow_parser_mgr.java:99)
        gplx.xowa.wikis.pages.Xowe_page_mgr.Load_page(Xowe_page_mgr.java:91)
        gplx.xowa.apps.servers.http.Http_server_mgr.Parse_page_to_html(Http_server_mgr.java:134)
        gplx.xowa.apps.servers.http.Http_server_wkr.Write_wiki(Http_server_wkr.java:111)
        gplx.xowa.apps.servers.http.Http_server_wkr.Process_get(Http_server_wkr.java:82)
        gplx.xowa.apps.servers.http.Http_server_wkr.Run(Http_server_wkr.java:60)
        gplx.xowa.apps.servers.http.Http_server_wkr.Invk(Http_server_wkr.java:283)
        gplx.Gfo_invk_.Invk_by_msg(Gfo_invk_.java:34)
        gplx.core.threads.Thread_adp.run(Thread_adp.java:33)
        java.lang.Thread.run(Thread.java:748)
I realise this is of my own making
I cannot work out the steps from Template:Azienda to Module:Controllo Wikidata and was hoping that being able to view the Lua stack might help

On the other hand it might not

**Comment by gnosygnu** [comment]
Not really. In the above case, Luaj (Java code) isn't throwing an exception. The exception is actually being thrown by itwiki:Module (Lua code) so the stack trace is up to Lua (which you sort of get: see below)

Even when the error is in Luaj code, the Java stack trace is rarely useful. Luaj uses a virtual machine to run the code and the stack trace is just interleaved with the same vm func over and over again.

To debug the above, I'd try something like the following:
* Go to https://it.wikipedia.org/wiki/Modulo:Wikidata&action=edit
* Find line 721 and put some debug statement
```
dbg("luaj_debug", entity.datatype);
return i18n.datatypes[entity.datatype]
```
* Go to https://it.wikipedia.org/wiki/Airbus_Group&action=edit and Preview it
* See what the output is of entity.datatype and go from there

It's sort of painful, so if you run into issues, let me know, and I'll pull down it.wikipedia.org later.

Also, on another topic, I'm guessing you can read some Italian? FYI: I have passing familiarity with French.

```
Modulo:Wikidata:721 Tipo di dato sconosciuto
  stack traceback:
  Module:Controllo Wikidata:32: in function <Module:Controllo Wikidata:31>
  mw.lua:531: in function <mw.lua:530>
  [Java]: in function '__index'
  g:/xowa/bin/any/xowa/xtns/Scribunto/engines/Luaj/MWServer.lua:75:
```

**Comment by desb42** [comment]
Did French at school (holiday there a lot) and worked in Italy and Germany
As yet have not downloaded frwiki

**Comment by gnosygnu** [comment]
Yeah, I took French for multiple years over high-school and college. Fairly rusty now though.

Will probably try frwiki sometime after itwiki.

Thanks.
